### ISSUE TITLE
Declaration emit broken when class method returns new instance

### ISSUE THREAD

**Issue by ethanresnick** [Microsoft/TypeScript#21958]
**TypeScript Version:**  2.7.0-dev.201xxxxx

**Search Terms:** 
declaration emit default broken constructor

**Code**

```ts
// index.ts
export default class Data<T> {
  protected constructor(private value: T[]) { }

  map<U>(fn: (it: T) => Data<U>) {
    return new Data(this.value.map(fn));
  }

  static of<U>(data: U[]) {
    return new Data<U>(data);
  }
}
```

```json
// tsconfig.json (in case it's relevant)
{
  "compilerOptions": {
    "lib": ["es6", "es2015.promise"],
    "module": "commonjs",
    "moduleResolution": "node",
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "outDir": "./build",
    "preserveConstEnums": true,
    "removeComments": true,
    "strict": true,
    "noImplicitAny": false,
    "target": "es6",
    "declaration": true
  }
}
```

**Expected behavior:**
The generated declarations should look like the below, which is the output using TS 2.6.2

```ts
// index.d.ts
export default class Data<T> {
    private value;
    protected constructor(value: T[]);
    map<U>(fn: (it: T) => Data<U>): Data<U>;
    static of<U>(data: U[]): Data<U>;
}
```

**Actual behavior:**
The declarations look like the below, which don't work when referenced from another package. Note the `default<U>` and `default<default<U>>`.

```ts
export default class Data<T> {
    private value;
    protected constructor(value: T[]);
    map<U>(fn: (it: T) => Data<U>): default<default<U>>;
    static of<U>(data: U[]): default<U>;
}
```

**Comment by mhegazy** [comment]
looks like a duplicate of https://github.com/Microsoft/TypeScript/issues/21520. should be fixed in ts 2.7.2

**Comment by ethanresnick** [comment]
Sounds good. I believe it was still happening for me in the latest 2.8.0-insiders, though, so if that build is supposed to include the fix for #21520, this may be something slightly different.

**Comment by mhegazy** [comment]
```sh
c:\test\21958>type a.ts
// index.ts
export default class Data<T> {
    protected constructor(private value: T[]) { }

    map<U>(fn: (it: T) => Data<U>) {
      return new Data(this.value.map(fn));
    }

    static of<U>(data: U[]) {
      return new Data<U>(data);
    }
  }
c:\test\21958>type tsconfig.json
// tsconfig.json (in case it's relevant)
{
    "compilerOptions": {
      "lib": ["es6", "es2015.promise"],
      "module": "commonjs",
      "moduleResolution": "node",
      "noImplicitReturns": true,
      "noUnusedLocals": true,
      "outDir": "./build",
      "preserveConstEnums": true,
      "removeComments": true,
      "strict": true,
      "noImplicitAny": false,
      "target": "es6",
      "declaration": true
    }
  }
c:\test\21958>tsc --v
Version 2.8.0-dev.20180215

c:\test\21958>tsc

c:\test\21958>type build\a.d.ts
export default class Data<T> {
    private value;
    protected constructor(value: T[]);
    map<U>(fn: (it: T) => Data<U>): Data<Data<U>>;
    static of<U>(data: U[]): Data<U>;
}

```
